### PR TITLE
Include request info in exceptions to make it easy to see what reques…

### PR DESCRIPTION
**Pull request by grosser**
…t and what server failed

we are doing a lot of requests to a lot of different servers, so a general "This failed" exception is not helpful.

```
Kubernetes::Cluster.last.client.get_secret 'sdfsdfs'
KubeException: HTTP status code 404 the server could not find the requested resource for GET https://k8s-sandbox.consul:8443/api/v1/secrets/sdfsdfs
  from kubeclient/common.rb:107:in `rescue in handle_exception'
```

@abonas

### CODE REVIEW

File: lib/kubeclient/common.rb [hunk 1]
```diff
@@ -104,7 +104,7 @@ def handle_exception
         {}
       end
       err_message = json_error_msg['message'] || e.message
-      raise KubeException.new(e.http_code, err_message, e.response)
+      fail KubeException.new(e.http_code, err_message, e.response)
```
Comment (grosser) [inline lib/kubeclient/common.rb]: making this consistent with the other usages

File: lib/kubeclient/kube_exception.rb [hunk 2]
```diff
@@ -9,6 +9,10 @@ def initialize(error_code, message, response)
   end
 
   def to_s
```
Comment (grosser) [inline lib/kubeclient/kube_exception.rb]: ideally I'd like to modify the `@message` in `initialize`, but that would be a bigger change ...

**Comment by simon3z** [comment]
@grosser just a couple of thoughts (to be discussed): could this additional information being reported in one (or more) separate attributes of KubeException? (For example it may be nice to keep the URL as an URL instead of stringify it).
Can we report all this info only when specifically requested? E.g. `ex.verbose_message`? (Naming is probably not good but you get the idea...)

@moolitayer @dkorn @enoodle @zeari please review.

**Comment by grosser** [comment]
I'd prefer more output by default ... it's easy to ignore, but hard to get
when something goes wrong and was not set up to be verbose
separate attributes are not really necessary since all that information is
already available via the request

File: lib/kubeclient/kube_exception.rb [hunk 3]
```diff
@@ -9,6 +9,10 @@ def initialize(error_code, message, response)
...
-    'HTTP status code ' + @error_code.to_s + ', ' + @message
+    string = "HTTP status code #{@error_code} #{@message}"
```
Comment (moolitayer) [inline lib/kubeclient/kube_exception.rb]: Removed the ',' on purpose?

File: lib/kubeclient/common.rb [hunk 4]
```diff
@@ -265,7 +265,7 @@ def get_entities(entity_type, klass, resource_name, options = {})
       resource_version = result.fetch('resourceVersion', nil)
       if resource_version.nil?
         resource_version =
-            result.fetch('metadata', {}).fetch('resourceVersion', nil)
+          result.fetch('metadata', {}).fetch('resourceVersion', nil)
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @grosser fixing the indentation style (here and below) is not related to this PR. Please send a separate PR (which I suppose will be merged much faster).

**Comment by grosser** [comment]
updated

**Comment by zeari** [comment]
LGTM @simon3z

File: lib/kubeclient/kube_exception.rb [hunk 5]
```diff
@@ -9,6 +9,10 @@ def initialize(error_code, message, response)
   end
 
   def to_s
-    'HTTP status code ' + @error_code.to_s + ', ' + @message
+    string = "HTTP status code #{@error_code}, #{@message}"
+    if @response.is_a?(RestClient::Response) && @response.request
+      string << " for #{@response.request.method.upcase} #{@response.request.url}"
```
Comment (simon3z) [inline lib/kubeclient/kube_exception.rb]: @grosser should we protect in case any of these are ever `nil`?

This is a pretty dangerous part where to get `NoMethodError` exceptions. cc @cben

Comment (grosser) [inline lib/kubeclient/kube_exception.rb]: checking that the request is there above ... and method can never be nil ... so I think this is safe

Comment (cben) [inline lib/kubeclient/kube_exception.rb]: LGTM.  RestClient::Request [normalizes method][1] to always be a string.

[1]: https://github.com/rest-client/rest-client/blob/f62a8a0e/lib/restclient/request.rb#L707-L710
